src/lib.rs

```diff
@@ -795,5 +886,252 @@ where F: FnOnce() -> R,
 
 #[test]
 fn simple_build() {
-    Builder::new().build().expect("Couldn't build default Counter");
+    Builder::new()
+        .counter()
+        .expect("Couldn't build default Counter");
+}
+
+// Use a pretty big buffer because we don't want to drop any entries
+const SAMPLE_BUFFER_PAGES: usize = 128 + 1;
+
+/// Size of a page of memory in bytes.
+fn page_size() -> usize {
+    (unsafe { libc::sysconf(libc::_SC_PAGESIZE) }) as usize
+}
+
+fn wait_for_readable_or_timeout(file: &File, timeout: Option<std::time::Duration>) -> bool {
+    let mut pollfd = pollfd {
+        fd: file.as_raw_fd(),
+        events: POLLIN,
+        revents: 0,
+    };
+    let timeout = timeout.map(|d| d.as_millis() as c_int).unwrap_or(-1);
+    let events = unsafe { poll(&mut pollfd, 1, timeout) };
+    events == 0
+}
+
+// We create our own version of this rather than use bindgen's so we can have the atomics.
+#[repr(C)]
+struct PerfEventMmapPage {
+    /// version number of this structure
+    version: u32,
+
+    /// lowest version this is compat with
+    compat_version: u32,
+
+    /// seqlock for synchronization
+    lock: u32,
+
+    /// hardware counter identifier
+    index: u32,
+
+    /// add to hardware counter value
+    offset: i64,
+
+    /// time event active
+    time_enabled: u64,
+
+    /// time event on CPU
+    time_running: u64,
+
+    capabilites: u64,
+    pmc_width: u16,
+    time_shift: u16,
+    time_mult: u32,
+    time_offset: u64,
+    __reserved: [u64; 120], /* Pad to 1 k */
+
+    /// head in the data section
+    data_head: AtomicU64,
+
+    /// user-space written tail
+    data_tail: AtomicU64,
```

Comment:
I wonder if casting `*const u64` pointers to `&AtomicU64` (or even having a `PerfEventDataBuffer` struct with the `data_*` fields) would be preferable to copying the `struct`.

Or maybe there's a way to have `bindgen` use different types for some fields in the `-sys` crate?

(I'm interested in this because it's adjacent to the `rdpmc` usage, although all examples I've seen for that just compare the value of `lock` before/after `rdpmc`, and use barriers, without any otherwise-synchronized loads)